File: src/main/java/com/github/daniel/shuy/kotlin/spring/boot/example/service/PetService.java
package com.github.daniel.shuy.kotlin.spring.boot.example.service;

import com.github.daniel.shuy.kotlin.spring.boot.example.dto.PetDto;
import com.github.daniel.shuy.kotlin.spring.boot.example.dto.PetFilterDto;
import com.github.daniel.shuy.kotlin.spring.boot.example.mapper.PetMapper;
import com.github.daniel.shuy.kotlin.spring.boot.example.model.Pet;
import com.github.daniel.shuy.kotlin.spring.boot.example.properties.PetProperties;
import com.github.daniel.shuy.kotlin.spring.boot.example.repository.PetRepository;
import com.github.daniel.shuy.kotlin.spring.boot.example.specification.PetSpecifications;
import java.util.Map;
import java.util.regex.Pattern;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
@RequiredArgsConstructor
@Slf4j
public class PetService {
  private static final Pattern REGEX_ALPHABETIC_OR_SPACE = Pattern.compile("[a-zA-Z ]*");

  private final PetMapper petMapper;
  private final PetProperties petProperties;
  private final PetRepository petRepository;
  private final PetSpecifications petSpecifications;

  public Pet createPet(PetDto petDto) {
    validateName(petDto);

    var pet = petMapper.mapToEntity(petDto);

    var savedPet = petRepository.save(pet);
    log.info("Created Pet with id {}", savedPet.getId());
    return savedPet;
  }

  public Page<Pet> findPets(PetFilterDto petFilterDto, Pageable pageable) {
    var specification = petFilterDto.toSpecification(petSpecifications);
    return petRepository.findAll(specification, pageable);
  }

  public Pet findPetById(Long petId) {
    return petRepository.findById(petId)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
  }

  public Pet updatePet(PetDto petDto) {
    var pet = findPetById(petDto.id());

    validateName(petDto);

    petMapper.updateEntityFromDto(petDto, pet);

    var savedPet = petRepository.save(pet);
    log.info("Updated Pet with id {}", savedPet.getId());
    return savedPet;
  }

  public void deletePet(Long petId) {
    var pet = findPetById(petId);

    petRepository.delete(pet);
    log.info("Deleted Pet with id {}", pet.getId());
  }

  public Map<String, Number> getSoldCountByTag() {
    return petRepository.getSoldCountByTag();
  }

  protected void validateName(PetDto petDto) {
    var name = petDto.name();
    if (name == null || name.isEmpty()) {
      return;
    }

    var nameMaxLength = petProperties.nameMaxLength();
    if (nameMaxLength != null && name.length() > nameMaxLength) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, String.format("Name cannot be longer than %d", nameMaxLength));
    }

    if (!REGEX_ALPHABETIC_OR_SPACE.matcher(name).matches()) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Name can only contain alphabetic or space characters");
    }
  }
}

File: src/main/kotlin/com/github/daniel/shuy/kotlin/spring/boot/example/service/PetService.kt
package com.github.daniel.shuy.kotlin.spring.boot.example.service

import com.github.daniel.shuy.kotlin.spring.boot.example.dto.PetDto
import com.github.daniel.shuy.kotlin.spring.boot.example.dto.PetFilterDto
import com.github.daniel.shuy.kotlin.spring.boot.example.mapper.PetMapper
import com.github.daniel.shuy.kotlin.spring.boot.example.model.Pet
import com.github.daniel.shuy.kotlin.spring.boot.example.properties.PetProperties
import com.github.daniel.shuy.kotlin.spring.boot.example.repository.PetRepository
import com.github.daniel.shuy.kotlin.spring.boot.example.specification.PetSpecifications
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException

private val logger = KotlinLogging.logger {}

@Service
class PetService(
    private val petMapper: PetMapper,
    private val petProperties: PetProperties,
    private val petRepository: PetRepository,
    private val petSpecifications: PetSpecifications,
) {
    companion object {
        private val REGEX_ALPHABETIC_OR_SPACE = Regex("[a-zA-Z ]*")
    }

    fun createPet(petDto: PetDto): Pet {
        validateName(petDto)
        val pet = petMapper.mapToEntity(petDto)
        return petRepository.save(pet).also {
            logger.info { "Created Pet with id ${it.id}" }
        }
    }

    fun findPets(petFilterDto: PetFilterDto, pageable: Pageable): Page<Pet> {
        val specification = petFilterDto.toSpecification(petSpecifications)
        return petRepository.findAll(specification, pageable)
    }

    fun findPetById(petId: Long): Pet {
        return petRepository.findByIdOrNull(petId)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
    }

    fun updatePet(petDto: PetDto): Pet {
        val petId = petDto.id
            ?: throw ResponseStatusException(HttpStatus.BAD_REQUEST, "id must not be null")
        val pet = findPetById(petId)
        validateName(petDto)
        petMapper.updateEntityFromDto(petDto, pet)
        return petRepository.save(pet).also {
            logger.info { "Updated Pet with id ${it.id}" }
        }
    }

    fun deletePet(petId: Long) {
        val pet = findPetById(petId)
        petRepository.delete(pet)
        logger.info { "Deleted Pet with id ${pet.id}" }
    }

    fun getSoldCountByTag(): Map<String, Number> {
        return petRepository.getSoldCountByTag()
    }

    protected fun validateName(petDto: PetDto) {
        val name = petDto.name
        if (name.isEmpty()) {
            return
        }

        petProperties.nameMaxLength?.let { nameMaxLength ->
            if (name.length > nameMaxLength) {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Name cannot be longer than $nameMaxLength")
            }
        }

        if (!REGEX_ALPHABETIC_OR_SPACE.matches(name)) {
            throw ResponseStatusException(
                HttpStatus.BAD_REQUEST,
                "Name can only contain alphabetic or space characters",
            )
        }
    }
}
